Form properties and methods (#226)
javascript-tutorial/it.javascript.info: 2-ui/4-forms-controls/1-form-elements/article.md

# Proprietà dei form e metodi

I forms e gli elementi control, come `<input>` hanno una serie di eventi e proprietà peculiari.

> Lavorando con i forms sarà molto più comodo quando li avremo imparati.

Reviewer: ```suggestion
Lavorando con i forms, questi saranno molto comodi quando li avremo imparati.
```

Author: /done

> ## Navigazione: form and elements

Reviewer: ```suggestion
## Navigazione: form e elements
```

Author: /done

I form del documento sono membri della speciale collezione `document.forms`.

Questa è una cosiddetta "named collection": è sia associativa che ordinata. Possiamo usare sia il nome che l'indice nel documento per accedervi.

```js no-beautify
document.forms.my - il form con name="my"
document.forms[0] - il primo form del documento
```

> Quando abbiamo un form, allora qualunque elemento dentro di esso sarà contenuto nella named collection `form.elements`.

Reviewer: ```suggestion
Quando abbiamo un form, allora tutti gli elementi saranno contenuti nella named collection `form.elements`.
```

Author: /done

    let fieldset = form.elements.userFields;
    alert(fieldset); // HTMLFieldSetElement

    // possiamo ottenere l'input sia dal nome del form sia dal fieldset
    alert(fieldset.elements.login == form.elements.login); // true
*/!*
  </script>
</body>
```
````

````warn header="Notazione breve: `form.name`"

> C'è una notazione breve: possiamo accedere all'elemento come `form[index/name]`.

Reviewer: ```suggestion
Esiste una notazione breve: possiamo accedere all'elemento come `form[index/name]`.
```

Author: /done

In altre parole, invece di `form.elements.login` possiamo scrivere `form.login`.

> Funziona ugualmente, ma c'è un piccolo problema: se accediamo a un elemento, e dopo cambia il suo  `name`, allora sarà ancora disponibile attraverso il vecchio nome (ma anche da quello nuovo).

Reviewer: ```suggestion
Funziona ugualmente, ma c'è un piccolo problema: se accediamo a un elemento, che in successivamente cambia il suo `name`, questo sarà ancora accessibile sia attraverso il vecchio nome, ma anche tramite quello nuovo.
```

Author: /done

È facile capirlo da un esempio:

```html run height=40
<form id="form">
  <input name="login">
</form>

<script>
  alert(form.elements.login == form.login); // true, lo stesso <input>

  form.login.name = "username"; // cambio del nome dell'input

  // form.elements ha aggiornato il nome:
  alert(form.elements.login); // undefined
  alert(form.elements.username); // input

*!*
  // form permette entrambi i nomi: sia quello nuovo che quello vecchio
  alert(form.username == form.login); // true
*/!*
</script>
```

Solitamente non è un problema, in quanto raramente andiamo a modificare il nome degli elementi dei form.

````

## Backreference: element.form

Per ogni elemento, il form è disponibile come `element.form`. Quindi un form referenzia tutti gli elementi, e gli elementi referenziano il form.

> Ecco un figura:

Reviewer: ```suggestion
Come possiamo vedere in figura:
```

Author: /done

</script>
```

## Elementi del form

Parliamo un po' dei form controls.

### input e textarea

Possiamo accedere ai lori valori tramite `input.value` (string) o `input.checked` (boolean) per i checkbox.

Come in questo caso:

```js
input.value = "Nuovo valore";
textarea.value = "Nuovo testo";

> input.checked = true; // for a checkbox o per un radio button

Reviewer: ```suggestion
input.checked = true; // per una checkbox o un radio button
```

Author: /done

```

```warn header="Usare `textarea.value`, e non `textarea.innerHTML`"

> Nota bene che sebbene anche `<textarea>...</textarea>` contenga il suo valore come HTML annidato, non dovremmo mai usare `textarea.innerHTML` per accedervi.

Reviewer: ```suggestion
Nota bene che, nonostante anche `<textarea>...</textarea>` contenga il suo valore come HTML annidato, non dovremmo mai usare `textarea.innerHTML` per accedervi.
```

Author: /done

2. `select.value` -- il *valore* di `<option>` attualmente selezionato,
3. `select.selectedIndex` -- l'*indice* di `<option>` attualmente selezionato.

Forniscono tre modi differenti per impostare un valore per un `<select>`:

1. Trova il corrispondente elemento `<option>` (ad esempio tra i `select.options`) ed imposta il suo `option.selected` a `true`.
2. Se conosciamo il nuovo valore: imposta `select.value` al nuovo valore.
3. Se conosciamo l'indice della nuova opzione: imposta `select.selectedIndex` su quell'indice.

Ecco un esempio per tutti e tre i metodi:

```html run
<select id="select">
  <option value="apple">Mela</option>
  <option value="pear">Pera</option>
  <option value="banana">Banana</option>
</select>

<script>
  // tutte e tre le righe di codice fanno la stessa cosa
  select.options[2].selected = true; 
  select.selectedIndex = 2;
  select.value = 'banana';
  // nota bene: le options cominciano da indice zero, quindi indice 2 significa la option numero 3.
</script>
```

Diversamente da altri controls, `<select>` permette più opzioni alla volta se contiene l'attributo `multiple`. Sebbene questo attributo venga usato raramente.

> Per valori multipli selezionati, usiamo il primo modo per impostare i valori: aggiungere/rimuovere la proprietà `selected` dai sottolementi `<option>`.

Reviewer: ```suggestion
Per valori multipli selezionati, usiamo il primo modo per impostare i valori: aggiungere/rimuovere la proprietà `selected` dai sottoelementi `<option>`.
```

Author: /done

</script>
```

Le specifiche complete dell'elemento `<select>` sono disponibili nelle specifiche <https://html.spec.whatwg.org/multipage/forms.html#the-select-element>.

### new Option

> Nelle specifiche [specification](https://html.spec.whatwg.org/multipage/forms.html#the-option-element) c'è una sintassi breve e carina per creare una elemento  `<option>`:

Reviewer: ```suggestion
Nelle specifiche [specification](https://html.spec.whatwg.org/multipage/forms.html#the-option-element) viene descritta una sintassi breve ed elegante per creare una elemento  `<option>`:
```

Author: /done

```js
option = new Option(text, value, defaultSelected, selected);
```

Questa sintassi è opzionale. Possiamo usare `document.createElement('option')` ed impostare gli attributi manualmente. Tuttavia, potrebbe essere breve, quindi ecco i parametri:

- `text` -- il testo dentro option,
- `value` -- il valore di option,
- `defaultSelected` -- se `true`, allora verrà creato l'attributo HTML `selected`,
- `selected` -- se `true`, allora l'opzione verrà selezionata.

> La differenza tra `defaultSelected` e `selected` è che `defaultSelected` imposta l'attributo HTML (che possiamo ottenere usando `option.getAttribute('selected')`, mentre `selected` decide se l'opzione è selezionata o meno.

Reviewer: ```suggestion
La differenza tra `defaultSelected` e `selected` è che `defaultSelected` imposta l'attributo HTML (che possiamo ottenere usando `option.getAttribute('selected')`, mentre `selected` definisce lo stato della selezione (se è selezionata o meno).
```

Author: /done

In pratica, solitamente possiamo impostare entrambi i valori a `true` o `false` (oppure ometterli, che equivale a `false`).

Per esempio, ecco un nuovo elemento option "non selezionato":

```js
let option = new Option("Testo", "value");
// crea <option value="value">Testo</option>
```

La stesso elemento option, ma stavolta selezionato:

```js
let option = new Option("Testo", "value", true, true);
```

Gli elementi option hanno delle proprietà:

`option.selected`
: Se l'opzione è selezionata.

`option.index`
: L'indice dell'opzione in mezzo agli altri elementi option del suo elemento `<select>`.

`option.text`
: Il contenuto testuale dell'elemento option (visto dall'utente).

## Riferimenti

- Specification: <https://html.spec.whatwg.org/multipage/forms.html>.

## Riepilogo

Navigazione dei form:

`document.forms`
: Un form è disponibile come `document.forms[name/index]`.

`form.elements`  

> : Gli elementi del form sono disponibili come `form.elements[name/index]`, oppure si può anche solo usare `form[name/index]`. La proprietà `elements` esiste anche per i `<fieldset>`.

Reviewer: ```suggestion
: Gli elementi del form sono disponibili come `form.elements[name/index]`, oppure, più semplicemente con `form[name/index]`. La proprietà `elements` esiste anche per i `<fieldset>`.
```

Author: /done

`element.form`
: Gli elementi referenziano i loro form nella proprietà `form`.

Il valore è disponibile come `input.value`, `textarea.value`, `select.value` etc, o come `input.checked` per i checkbox e radio buttons.

Per `<select>` possiamo anche ottenere il valore tramite l'indice `select.selectedIndex` o attraverso la collezione di options `select.options`.

> Questi sono le basi da cui partire con i form. Incontreremo molti esempi più avanti nel tutorial.

Reviewer: ```suggestion
Queste sono le basi da cui partire con i form. Incontreremo molti esempi più avanti nel tutorial.
```

Author: /done
